include/rmm/device_scalar.hpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/device_buffer.hpp>
+#include <rmm/mr/default_memory_resource.hpp>
+#include <rmm/mr/device_memory_resource.hpp>
+
+namespace rmm {
+
+/**---------------------------------------------------------------------------*
+ * @brief Container for a single object of type `T` in device memory.
+ * 
+ * `T` must be trivially copyable.
+ *
+ * @tparam T The object's type
+ *---------------------------------------------------------------------------**/
+template <typename T>
+class device_scalar {
+ public:
+  static_assert(std::is_trivially_copyable<T>::value,
+                "Scalar type must be trivially copyable");
+
+  /**---------------------------------------------------------------------------*
+   * @brief Construct a new `device_scalar`
+   *
+   * @param initial_value The initial value of the object in device memory
+   * @param stream_ Optional, stream on which to perform allocation and copy
+   * @param mr_ Optional, resource with which to allocate
+   *---------------------------------------------------------------------------**/
+  explicit device_scalar(
+      T initial_value, cudaStream_t stream_ = 0,
+      rmm::mr::device_memory_resource *mr_ = rmm::mr::get_default_resource())
+      : buff{sizeof(T), stream_, mr_} {
+    auto status = cudaMemcpyAsync(buff.data(), &initial_value, sizeof(T),
+                                  cudaMemcpyDefault, buff.stream());
+
+    if (cudaSuccess != status) {
+      throw std::runtime_error{"Device memcpy failed."};
+    }
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Synchronizes the `device_scalar`'s stream and copies the value to
+   * the host.
+   *
+   * @return T The value of the scalar after synchronizing its stream
+   *---------------------------------------------------------------------------**/
+  T value() const noexcept {
+    T host_value{};
+    auto status = cudaStreamSynchronize(buff.stream());
```

Comment:
Need to synchronize after the call to cudaMemcpyAsync in order to return the host value correctly. Why synchronize before copying?

Reply:
Blah, absent minded.